lib/Optimizer/Dialect/FIRType.cpp

```diff
@@ -0,0 +1,1288 @@
+    if (finalized)
+      return;
+    finalized = true;
+    setLenParamList(lenParamList);
+    setTypeList(typeList);
+  }
+
+protected:
+  std::string name;
+  bool finalized;
+  std::vector<RecordType::TypePair> lens;
+  std::vector<RecordType::TypePair> types;
+
+private:
+  RecordTypeStorage() = delete;
+  explicit RecordTypeStorage(llvm::StringRef name)
+      : name{name}, finalized{false} {}
+};
+
+/// Type descriptor type storage
+struct TypeDescTypeStorage : public mlir::TypeStorage {
+  using KeyTy = mlir::Type;
+
+  static unsigned hashKey(const KeyTy &key) { return llvm::hash_combine(key); }
+
+  bool operator==(const KeyTy &key) const { return key == getOfType(); }
+
+  static TypeDescTypeStorage *construct(mlir::TypeStorageAllocator &allocator,
+                                        mlir::Type ofTy) {
+    assert(ofTy && "descriptor type is null");
+    auto *storage = allocator.allocate<TypeDescTypeStorage>();
+    return new (storage) TypeDescTypeStorage{ofTy};
+  }
+
+  // The type described by this type descriptor instance
+  mlir::Type getOfType() const { return ofTy; }
+
+protected:
+  mlir::Type ofTy;
+
+private:
+  TypeDescTypeStorage() = delete;
+  explicit TypeDescTypeStorage(mlir::Type ofTy) : ofTy{ofTy} {}
+};
+
+} // namespace detail
+
+template <typename A, typename B>
+bool inbounds(A v, B lb, B ub) {
+  return v >= lb && v < ub;
+}
+
+bool isa_fir_type(mlir::Type t) {
+  return inbounds(t.getKind(), mlir::Type::FIRST_FIR_TYPE,
+                  mlir::Type::LAST_FIR_TYPE);
+}
+
+bool isa_std_type(mlir::Type t) {
+  return inbounds(t.getKind(), mlir::Type::FIRST_STANDARD_TYPE,
+                  mlir::Type::LAST_STANDARD_TYPE);
+}
+
+bool isa_fir_or_std_type(mlir::Type t) {
+  return isa_fir_type(t) || isa_std_type(t);
+}
+
+bool isaMemref(mlir::Type t) {
+  return t.isa<ReferenceType>() || t.isa<PointerType>() || t.isa<HeapType>();
+}
+
+bool isanAggregate(mlir::Type t) {
+  return t.isa<SequenceType>() || t.isa<RecordType>();
+}
+
+mlir::Type dyn_cast_ptrEleTy(mlir::Type t) {
+  if (auto p = t.dyn_cast<fir::ReferenceType>())
+    return p.getEleTy();
+  if (auto p = t.dyn_cast<fir::PointerType>())
+    return p.getEleTy();
+  if (auto p = t.dyn_cast<fir::HeapType>())
+    return p.getEleTy();
+  return {};
+}
+
+} // namespace fir
+
+// CHARACTER
+
+CharacterType fir::CharacterType::get(mlir::MLIRContext *ctxt, KindTy kind) {
+  return Base::get(ctxt, FIR_CHARACTER, kind);
+}
+
+int fir::CharacterType::getFKind() const { return getImpl()->getFKind(); }
+
+// Dims
+
+DimsType fir::DimsType::get(mlir::MLIRContext *ctxt, unsigned rank) {
+  return Base::get(ctxt, FIR_DIMS, rank);
+}
+
+int fir::DimsType::getRank() const { return getImpl()->getRank(); }
+
+// Field
+
+FieldType fir::FieldType::get(mlir::MLIRContext *ctxt, KindTy) {
+  return Base::get(ctxt, FIR_FIELD, 0);
+}
+
+// Len
+
+LenType fir::LenType::get(mlir::MLIRContext *ctxt, KindTy) {
+  return Base::get(ctxt, FIR_LEN, 0);
+}
+
+// LOGICAL
```

Comment:
Shameless plug, but I've found the following comment block style to make chunking files much more readable:
https://github.com/llvm/llvm-project/blob/c3595d1069277b4ab0df49d7139b6f1bbc94f21c/mlir/lib/IR/Block.cpp#L15

Reply:
Noting this for a later PR.